test/test_connectionpool.py

```diff
@@ -226,6 +226,45 @@ def test_contextmanager(self):
 
         self.assertRaises(Empty, old_pool_queue.get, block=False)
 
+    def test_connection_recycling(self):
+        '''
+        Test that connections are recycled to the pool on 
+        connection/readtimeout errors where no http response 
+        is received.
+        '''
+        poolsize = 2
+        with HTTPConnectionPool('google.com', maxsize=poolsize, 
+                                  retries=1, block=True) as pool:
+
+            self.assertEqual(pool.pool.qsize(), poolsize)
+
+            # set a silly read timeout to force timeout exception. 
+            # We won't get a response for this and so the conn won't 
+            # be implicitly returned to the pool.
+            pool.timeout._read = 0
+            self.assertRaises(MaxRetryError, pool.urlopen, 'GET', 
+                              'http://google.com',
+                              release_conn=False)
+
+            # the pool should still contain poolsize elements
+            self.assertEqual(pool.pool.qsize(), poolsize)
+
+        with HTTPConnectionPool('localhost', maxsize=poolsize, 
+                                  retries=1, block=True) as pool:
+
+            self.assertEqual(pool.pool.qsize(), poolsize)
+
+            # force a connection error by supplying a 
+            # non-existent url. We won't get a response for this 
+            # and so the conn won't be implicitly returned to 
+            # the pool.
+            self.assertRaises(MaxRetryError, pool.urlopen, 'GET', 
+                              'localhost/nosuchurl',
+                              release_conn=False)
+
+            # the pool should still contain poolsize elements
+            self.assertEqual(pool.pool.qsize(), poolsize)
```

Comment:
Yes they force different exceptions, but actually they are handled in the same handler (for now) so we could just remove one of these. They are doubled up because they are slightly different and use a different host. You decide - create separate tests for each or remove one (which one - localhost or google)?